fix_: mitigate permission stuck in pending state

kounkou opened the pull request:
### Description

This PR mitigates permission stuck in pending state upon making device a control node. It fixes [#14023](https://github.com/status-im/status-desktop/issues/14023) by applying events and publishing community events once the device is promoted to control node.

### System Design

The main design decision for this issue is that promoteSelfToControlNode is inhibited when the community doesn't have a minted Owner token. From the UI perspective, the action is disabled similar to when the user is an `admin` user. And from the backend perspective, the action is also rejected with missing owner token code.

![Screenshot from 2024-05-09 15-59-48](https://github.com/status-im/status-go/assets/2589171/6e9dd4be-b6ee-42ac-a28f-0f8cb348d186)

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (83)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 44f42123bf8e8f05af42b6cc9adb828df7b5d2dc | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/1/) | 2024-04-18 08:34:56 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240418-083032-44f42123bf8e8f05af42b6cc9adb828df7b5d2dc-pr5070.zip) |
| :heavy_check_mark: | 44f42123bf8e8f05af42b6cc9adb828df7b5d2dc | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/1/) | 2024-04-18 08:36:06 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240418-083027-44f42123bf8e8f05af42b6cc9adb828df7b5d2dc-pr5070.aar) |
| :heavy_check_mark: | 44f42123bf8e8f05af42b6cc9adb828df7b5d2dc | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/1/) | 2024-04-18 08:36:13 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240418-083032-44f42123bf8e8f05af42b6cc9adb828df7b5d2dc-pr5070.zip) |
| :heavy_check_mark: | 44f42123bf8e8f05af42b6cc9adb828df7b5d2dc | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/1/) | 2024-04-18 09:15:43 | ~45 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 67327a67dda709a20defd95f8fd95c685286174e | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/2/) | 2024-04-18 18:58:01 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240418-185525-67327a67dda709a20defd95f8fd95c685286174e-pr5070.zip) |
| :heavy_check_mark: | 67327a67dda709a20defd95f8fd95c685286174e | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/2/) | 2024-04-18 18:58:48 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240418-185525-67327a67dda709a20defd95f8fd95c685286174e-pr5070.zip) |
| :heavy_check_mark: | 67327a67dda709a20defd95f8fd95c685286174e | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/2/) | 2024-04-18 19:01:00 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240418-185522-67327a67dda709a20defd95f8fd95c685286174e-pr5070.aar) |
| :heavy_check_mark: | 67327a67dda709a20defd95f8fd95c685286174e | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/2/) | 2024-04-18 19:36:59 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | cd10710a7e8495ae38d699bbf98ddd1ad8d289e5 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/3/) | 2024-04-19 08:01:20 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240419-075936-cd10710a7e8495ae38d699bbf98ddd1ad8d289e5-pr5070.aar) |
| :heavy_check_mark: | cd10710a7e8495ae38d699bbf98ddd1ad8d289e5 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/3/) | 2024-04-19 08:01:51 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240419-075936-cd10710a7e8495ae38d699bbf98ddd1ad8d289e5-pr5070.zip) |
| :heavy_check_mark: | cd10710a7e8495ae38d699bbf98ddd1ad8d289e5 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/3/) | 2024-04-19 08:04:33 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240419-075936-cd10710a7e8495ae38d699bbf98ddd1ad8d289e5-pr5070.zip) |
| :heavy_multiplication_x: | cd10710a7e8495ae38d699bbf98ddd1ad8d289e5 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/3/) | 2024-04-19 08:34:58 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/3//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | fab6ab360b1555b9827ba8cbb727c5dfea95d0f0 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/4/) | 2024-04-24 12:44:57 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/4//consoleText) |
| :heavy_check_mark: | fab6ab360b1555b9827ba8cbb727c5dfea95d0f0 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/4/) | 2024-04-24 12:47:19 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240424-124327-fab6ab360b1555b9827ba8cbb727c5dfea95d0f0-pr5070.zip) |
| :heavy_check_mark: | fab6ab360b1555b9827ba8cbb727c5dfea95d0f0 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/4/) | 2024-04-24 12:47:38 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240424-124327-fab6ab360b1555b9827ba8cbb727c5dfea95d0f0-pr5070.zip) |
| :heavy_check_mark: | fab6ab360b1555b9827ba8cbb727c5dfea95d0f0 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/4/) | 2024-04-24 12:48:07 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240424-124323-fab6ab360b1555b9827ba8cbb727c5dfea95d0f0-pr5070.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 962a8d9ef00c341144a746c9126ed894ab6fb477 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/5/) | 2024-04-25 02:08:56 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/5//consoleText) |
| :heavy_check_mark: | 962a8d9ef00c341144a746c9126ed894ab6fb477 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/5/) | 2024-04-25 02:11:17 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-020717-962a8d9ef00c341144a746c9126ed894ab6fb477-pr5070.zip) |
| :heavy_check_mark: | 962a8d9ef00c341144a746c9126ed894ab6fb477 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/5/) | 2024-04-25 02:11:30 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-020717-962a8d9ef00c341144a746c9126ed894ab6fb477-pr5070.zip) |
| :heavy_check_mark: | 962a8d9ef00c341144a746c9126ed894ab6fb477 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/5/) | 2024-04-25 02:12:20 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-020717-962a8d9ef00c341144a746c9126ed894ab6fb477-pr5070.aar) |
| | | | | | | |
| :heavy_multiplication_x: | acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/6/) | 2024-04-25 02:10:43 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/6//consoleText) |
| :heavy_check_mark: | acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/6/) | 2024-04-25 02:14:09 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-021133-acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2-pr5070.zip) |
| :heavy_check_mark: | acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/6/) | 2024-04-25 02:15:11 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-021235-acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2-pr5070.aar) |
| :heavy_check_mark: | acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/6/) | 2024-04-25 02:15:40 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-021150-acc4d2dfd15ebad1c88b2a7abbb6c5e1841b94c2-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | f184645033bd631f3a7b952e140e8ea5079b7d7a | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/7/) | 2024-04-25 02:26:43 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/7//consoleText) |
| :heavy_check_mark: | f184645033bd631f3a7b952e140e8ea5079b7d7a | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/7/) | 2024-04-25 02:27:45 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-022526-f184645033bd631f3a7b952e140e8ea5079b7d7a-pr5070.zip) |
| :heavy_check_mark: | f184645033bd631f3a7b952e140e8ea5079b7d7a | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/7/) | 2024-04-25 02:27:51 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-022522-f184645033bd631f3a7b952e140e8ea5079b7d7a-pr5070.aar) |
| :heavy_check_mark: | f184645033bd631f3a7b952e140e8ea5079b7d7a | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/7/) | 2024-04-25 02:29:12 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-022522-f184645033bd631f3a7b952e140e8ea5079b7d7a-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | e1aa5c1ec0ee9954f42834f2580b411e9719bab6 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/8/) | 2024-04-25 02:28:09 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/8//consoleText) |
| :heavy_check_mark: | e1aa5c1ec0ee9954f42834f2580b411e9719bab6 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/8/) | 2024-04-25 02:29:46 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-022806-e1aa5c1ec0ee9954f42834f2580b411e9719bab6-pr5070.aar) |
| :heavy_check_mark: | e1aa5c1ec0ee9954f42834f2580b411e9719bab6 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/8/) | 2024-04-25 02:30:09 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-022801-e1aa5c1ec0ee9954f42834f2580b411e9719bab6-pr5070.zip) |
| :heavy_check_mark: | e1aa5c1ec0ee9954f42834f2580b411e9719bab6 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/8/) | 2024-04-25 02:33:24 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-022937-e1aa5c1ec0ee9954f42834f2580b411e9719bab6-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | b1dfd4cc96708ad9c327018e83acc3372dae0a6c | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/9/) | 2024-04-25 02:33:11 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/9//consoleText) |
| :heavy_check_mark: | b1dfd4cc96708ad9c327018e83acc3372dae0a6c | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/9/) | 2024-04-25 02:33:47 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-023147-b1dfd4cc96708ad9c327018e83acc3372dae0a6c-pr5070.aar) |
| :heavy_check_mark: | b1dfd4cc96708ad9c327018e83acc3372dae0a6c | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/9/) | 2024-04-25 02:34:22 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-023147-b1dfd4cc96708ad9c327018e83acc3372dae0a6c-pr5070.zip) |
| :heavy_check_mark: | b1dfd4cc96708ad9c327018e83acc3372dae0a6c | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/9/) | 2024-04-25 02:37:12 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-023346-b1dfd4cc96708ad9c327018e83acc3372dae0a6c-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 9d35adee1a1d68c84172a01651a50ce2ed1c337b | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/10/) | 2024-04-25 09:51:12 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/10//consoleText) |
| :heavy_check_mark: | 9d35adee1a1d68c84172a01651a50ce2ed1c337b | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/10/) | 2024-04-25 09:52:11 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-094944-9d35adee1a1d68c84172a01651a50ce2ed1c337b-pr5070.aar) |
| :heavy_check_mark: | 9d35adee1a1d68c84172a01651a50ce2ed1c337b | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/10/) | 2024-04-25 09:52:16 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-094944-9d35adee1a1d68c84172a01651a50ce2ed1c337b-pr5070.zip) |
| :heavy_check_mark: | 9d35adee1a1d68c84172a01651a50ce2ed1c337b | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/10/) | 2024-04-25 09:53:26 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-094944-9d35adee1a1d68c84172a01651a50ce2ed1c337b-pr5070.zip) |
| | | | | | | |
| :heavy_check_mark: | 6fe326bd6f8c9326f317800f7228c19814b4a799 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/11/) | 2024-04-25 16:12:15 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240425-161037-6fe326bd6f8c9326f317800f7228c19814b4a799-pr5070.aar) |
| :heavy_check_mark: | 6fe326bd6f8c9326f317800f7228c19814b4a799 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/11/) | 2024-04-25 16:12:55 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240425-161037-6fe326bd6f8c9326f317800f7228c19814b4a799-pr5070.zip) |
| :heavy_check_mark: | 6fe326bd6f8c9326f317800f7228c19814b4a799 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/11/) | 2024-04-25 16:13:57 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240425-161037-6fe326bd6f8c9326f317800f7228c19814b4a799-pr5070.zip) |
| :heavy_multiplication_x: | 6fe326bd6f8c9326f317800f7228c19814b4a799 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/11/) | 2024-04-25 16:55:48 | ~45 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/11//consoleText) |
| :heavy_multiplication_x: | 6fe326bd6f8c9326f317800f7228c19814b4a799 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/12/) | 2024-04-25 21:57:12 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/12//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 212eebc1f96013a86ab6cb2ebdc0471f443cf53d | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/12/) | 2024-04-26 10:39:50 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240426-103614-212eebc1f96013a86ab6cb2ebdc0471f443cf53d-pr5070.zip) |
| :heavy_check_mark: | 212eebc1f96013a86ab6cb2ebdc0471f443cf53d | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/12/) | 2024-04-26 10:40:48 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240426-103615-212eebc1f96013a86ab6cb2ebdc0471f443cf53d-pr5070.zip) |
| :heavy_check_mark: | 212eebc1f96013a86ab6cb2ebdc0471f443cf53d | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/12/) | 2024-04-26 10:42:06 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240426-103614-212eebc1f96013a86ab6cb2ebdc0471f443cf53d-pr5070.aar) |
| :heavy_multiplication_x: | 212eebc1f96013a86ab6cb2ebdc0471f443cf53d | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/13/) | 2024-04-26 11:16:24 | ~40 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/13//consoleText) |
| :heavy_multiplication_x: | 212eebc1f96013a86ab6cb2ebdc0471f443cf53d | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/14/) | 2024-04-29 08:28:41 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/14//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 04dd0f89afa8266e5c966c15d54b3498946ea40a | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/15/) | 2024-04-30 08:07:29 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/15//consoleText) |
| :heavy_check_mark: | 04dd0f89afa8266e5c966c15d54b3498946ea40a | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/13/) | 2024-04-30 08:07:47 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240430-080600-04dd0f89afa8266e5c966c15d54b3498946ea40a-pr5070.aar) |
| :heavy_check_mark: | 04dd0f89afa8266e5c966c15d54b3498946ea40a | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/13/) | 2024-04-30 08:08:28 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240430-080604-04dd0f89afa8266e5c966c15d54b3498946ea40a-pr5070.zip) |
| :heavy_check_mark: | 04dd0f89afa8266e5c966c15d54b3498946ea40a | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/13/) | 2024-04-30 08:09:44 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240430-080604-04dd0f89afa8266e5c966c15d54b3498946ea40a-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 407fbd32 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/16/) | 2024-05-08 20:26:04 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/16//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 407fbd32732a396fe611a87c9466ead72141e99c | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/14/) | 2024-05-08 20:27:15 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240508-202317-407fbd32732a396fe611a87c9466ead72141e99c-pr5070.zip) |
| :heavy_check_mark: | 407fbd32732a396fe611a87c9466ead72141e99c | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/14/) | 2024-05-08 20:28:51 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240508-202317-407fbd32732a396fe611a87c9466ead72141e99c-pr5070.zip) |
| :heavy_check_mark: | 407fbd32732a396fe611a87c9466ead72141e99c | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/14/) | 2024-05-08 20:29:18 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240508-202317-407fbd32732a396fe611a87c9466ead72141e99c-pr5070.aar) |
| | | | | | | |
| :heavy_check_mark: | d9170f043f0e038a0ce376c86fd813aa7646ad67 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/15/) | 2024-05-08 20:36:09 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240508-203352-d9170f043f0e038a0ce376c86fd813aa7646ad67-pr5070.zip) |
| :heavy_check_mark: | d9170f043f0e038a0ce376c86fd813aa7646ad67 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/15/) | 2024-05-08 20:37:31 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240508-203352-d9170f043f0e038a0ce376c86fd813aa7646ad67-pr5070.zip) |
| :heavy_check_mark: | d9170f043f0e038a0ce376c86fd813aa7646ad67 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/15/) | 2024-05-08 20:39:25 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240508-203352-d9170f043f0e038a0ce376c86fd813aa7646ad67-pr5070.aar) |
| | | | | | | |
| :heavy_check_mark: | d9170f04 | [#17](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/17/) | 2024-05-08 21:15:38 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/17//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 3dde1e8b4c031bf617321a959e5305096809ff18 | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/16/) | 2024-05-08 23:58:23 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240508-235459-3dde1e8b4c031bf617321a959e5305096809ff18-pr5070.zip) |
| :heavy_check_mark: | 3dde1e8b4c031bf617321a959e5305096809ff18 | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/16/) | 2024-05-08 23:58:53 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240508-235459-3dde1e8b4c031bf617321a959e5305096809ff18-pr5070.zip) |
| :heavy_check_mark: | 3dde1e8b4c031bf617321a959e5305096809ff18 | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/16/) | 2024-05-09 00:00:41 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240508-235459-3dde1e8b4c031bf617321a959e5305096809ff18-pr5070.aar) |
| | | | | | | |
| :heavy_check_mark: | 3dde1e8b | [#18](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/18/) | 2024-05-09 00:38:05 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/18//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 295db89771dbe731143329b0f1c5fae0eaf7af21 | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/17/) | 2024-05-09 08:11:14 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240509-080849-295db89771dbe731143329b0f1c5fae0eaf7af21-pr5070.aar) |
| :heavy_check_mark: | 295db89771dbe731143329b0f1c5fae0eaf7af21 | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/17/) | 2024-05-09 08:11:24 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240509-080849-295db89771dbe731143329b0f1c5fae0eaf7af21-pr5070.zip) |
| :heavy_check_mark: | 295db89771dbe731143329b0f1c5fae0eaf7af21 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/17/) | 2024-05-09 08:12:29 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240509-080849-295db89771dbe731143329b0f1c5fae0eaf7af21-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 295db897 | [#19](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/19/) | 2024-05-09 08:43:58 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/19//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 8499cabee6dffb593e38d0fdbf8560754cc778ac | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/18/) | 2024-05-09 18:37:35 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240509-183502-8499cabee6dffb593e38d0fdbf8560754cc778ac-pr5070.zip) |
| :heavy_check_mark: | 8499cabee6dffb593e38d0fdbf8560754cc778ac | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/18/) | 2024-05-09 18:37:42 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240509-183502-8499cabee6dffb593e38d0fdbf8560754cc778ac-pr5070.aar) |
| :heavy_check_mark: | 8499cabee6dffb593e38d0fdbf8560754cc778ac | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/18/) | 2024-05-09 18:38:30 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240509-183502-8499cabee6dffb593e38d0fdbf8560754cc778ac-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 8499cabe | [#20](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/20/) | 2024-05-09 19:09:56 | ~34 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/20//consoleText) |
| :heavy_multiplication_x: | 8499cabe | [#21](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/21/) | 2024-05-09 19:52:31 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/21//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42 | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/19/) | 2024-05-09 20:18:14 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240509-201551-7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42-pr5070.zip) |
| :heavy_check_mark: | 7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42 | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/19/) | 2024-05-09 20:18:29 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240509-201548-7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42-pr5070.aar) |
| :heavy_check_mark: | 7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42 | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/19/) | 2024-05-09 20:19:20 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240509-201548-7e72ab1a59971b20e4cf44d7b12ef31ed3d42c42-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 7e72ab1a | [#22](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/22/) | 2024-05-09 20:50:52 | ~34 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/22//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 51cb49ec1f3bfb3892840555a2386d00874645e2 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/20/) | 2024-05-09 21:03:24 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240509-210053-51cb49ec1f3bfb3892840555a2386d00874645e2-pr5070.aar) |
| :heavy_check_mark: | 51cb49ec1f3bfb3892840555a2386d00874645e2 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/20/) | 2024-05-09 21:03:51 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240509-210053-51cb49ec1f3bfb3892840555a2386d00874645e2-pr5070.zip) |
| :heavy_check_mark: | 51cb49ec1f3bfb3892840555a2386d00874645e2 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/20/) | 2024-05-09 21:04:21 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240509-210053-51cb49ec1f3bfb3892840555a2386d00874645e2-pr5070.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 51cb49ec | [#23](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/23/) | 2024-05-09 21:36:09 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/23//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 7d8f81971514fa702c232bfd31f3cce4d2bce7cc | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5070/21/) | 2024-05-09 22:12:49 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240509-221009-7d8f81971514fa702c232bfd31f3cce4d2bce7cc-pr5070.zip) |
| :heavy_check_mark: | 7d8f81971514fa702c232bfd31f3cce4d2bce7cc | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5070/21/) | 2024-05-09 22:12:49 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240509-221009-7d8f81971514fa702c232bfd31f3cce4d2bce7cc-pr5070.aar) |
| :heavy_check_mark: | 7d8f81971514fa702c232bfd31f3cce4d2bce7cc | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5070/21/) | 2024-05-09 22:13:24 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240509-221009-7d8f81971514fa702c232bfd31f3cce4d2bce7cc-pr5070.zip) |
| | | | | | | |
| :heavy_check_mark: | 7d8f8197 | [#24](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/24/) | 2024-05-09 22:52:42 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5070/24//consoleText) |

kounkou commented on protocol/communities_events_utils_test.go:
```diff
@@ -2523,3 +2523,33 @@ func testBanMemberWithDeletingAllMessages(base CommunityEventsTestsInterface, co
+	s.Require().NotNil(controlNode)
+
+	response, err = sender.PromoteSelfToControlNode(community.ID())
+	s.Require().NoError(err)
```
Currently the code doesn't look complete, because `Error` obtained is : 

```bash
--- FAIL: TestAdminCommunityEventsSuite (4.16s)
    --- FAIL: TestAdminCommunityEventsSuite/TestPermissionAppliedUponMakingDeviceControlNode (4.16s)
        /home/jacques/Documents/work/status-go/protocol/communities_events_utils_test.go:2551: 
            	Error Trace:	/home/jacques/Documents/work/status-go/protocol/communities_events_utils_test.go:2551
            	            				/home/jacques/Documents/work/status-go/protocol/communities_messenger_admin_test.go:366
            	Error:      	Received unexpected error:
            	            	CommunityDescription was not signed by an owner
            	Test:       	TestAdminCommunityEventsSuite/TestPermissionAppliedUponMakingDeviceControlNode
FAIL
exit status 1
FAIL	github.com/status-im/status-go/protocol	4.200s
```

In the newly code I added, this error happens if the processEvents failed as shown below.

```go
    err = community.processEvents(message, appliedEvents)
    if err != nil {
	    return false, err
    }
```

osmaczko replied:
That's because `sender` is an admin not the owner. Only owner (to be more precise: one of its devices - aka control node) can update&sign the description.

osmaczko commented on protocol/communities/manager.go:
```diff
@@ -5188,6 +5199,35 @@ func (m *Manager) promoteSelfToControlNode(community *Community, clock uint64) (
 	}
 
+	if community.config.EventsData != nil {
+		appliedEvents := m.applyCommunityEvents(community)
```
I don't think that's needed. Events are applied in `community.processEvents` already. Applied events should be obtained after processing, same as in `HandleCommunityEventsMessage`: 

```golang
		appliedEvents := map[string]uint64{}
		for _, event := range community.config.EventsData.Events {
			appliedEvents[event.EventTypeID()] = event.CommunityEventClock
		}
```

osmaczko commented on protocol/communities/manager.go:
```diff
@@ -5188,6 +5199,35 @@ func (m *Manager) promoteSelfToControlNode(community *Community, clock uint64) (
+		if err != nil {
+			return false, err
+		}
+
```
I believe `m.handleAdditionalAdminChanges(community)` should be called here as well.

If possible, I would try to extract events appliance logic from `HandleCommunityEventsMessage` and use it both here and there.

kounkou replied:
Indeed, discussed offline. I have added `m.handleAdditionalAdminChanges(community)`. But I am still facing issues with making the code work.

osmaczko reviewed (commented):
To reproduce the issue I would base on `TestControlNodeUpdateSigner`. I'll try to come up with a test scaffold for you today.

osmaczko commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +734,135 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+	s.advertiseCommunityTo(s.john, community, s.bob)
+	s.john = nil
+
+	err = tt.RetryWithBackOff(func() error {
```
Please use `WaitOnMessengerResponse`

osmaczko replied:
`err` is not used

osmaczko commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +734,135 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+	s.Require().Equal(communities.TokenPermissionAdditionPending, addedPermission.State)
+
+	// Retrieve community link & community
+	err = tt.RetryWithBackOff(func() error {
```
Please use `WaitOnMessengerResponse`

kounkou replied:
This is not needed anymore

osmaczko commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +734,135 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
 	}
 }
+
+func (s *MessengerCommunitiesSignersSuite) TestPApplyendingCommunityEventsUponMakingDeviceControlNode() {
```
```suggestion
func (s *MessengerCommunitiesSignersSuite) TestApplyPendingCommunityEventsUponMakingDeviceControlNode() {
```

osmaczko commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +734,135 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+
+	grantPermission(&s.Suite, community, s.john, s.alice, protobuf.CommunityMember_ROLE_ADMIN)
+	s.advertiseCommunityTo(s.john, community, s.bob)
+	s.john = nil
```
Please ensure `john` is shutdown properly.

kounkou replied:
Indeed thanks. Addding the shutdown() call.

osmaczko commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +734,135 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+	})
+	s.Require().NoError(err)
+
+	s.collectiblesServiceMock.SetSignerPubkeyForCommunity(community.ID(), common.PubkeyToHex(&s.alice.identity.PublicKey))
```
This looks wrong. Why alice is the signer now? :thinking:

kounkou replied:
this is a mistake, the other device from John is the one supposed to be set control node. Fixed

osmaczko commented on protocol/communities/manager.go:
```diff
@@ -5188,11 +5189,68 @@ func (m *Manager) promoteSelfToControlNode(community *Community, clock uint64) (
 	return ownerChanged, nil
 }
 
+func (m *Manager) handleCommunityEvents(community *Community) error {
```
It would be great to reuse this function in `HandleCommunityEventsMessage`, otherwise we'll have logic duplicated.

kounkou replied:
reused the function, this update doesn't change the current implementation behavior.

osmaczko commented on protocol/communities/manager.go:
```diff
@@ -5188,11 +5189,68 @@ func (m *Manager) promoteSelfToControlNode(community *Community, clock uint64) (
 		return false, err
 	}
 
+	if community.IsControlNode() {
```
This check seems to defensive - it is in `promoteSelfToControlNode`.

osmaczko reviewed (commented):
Thanks, other then some minor quality improvements, looks good.

@mprakhov can you please have a look?

kounkou commented:
Hi there,
I am changing my approach here to get this issue fixed ASAP.
I will just rely on the QA testing for now, and I will ignore the integration testing for now.

mprakhov commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +732,127 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+		},
+		"no community update with accepted request",
+	)
+	s.Require().NoError(err)
```
here you need to check that johnsOtherDevice has a permission and permission in a pending state

kounkou replied:
Updated the integration tests. This is not relevant anymore

mprakhov commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +732,127 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
+
+	addedPermission := func() *communities.CommunityTokenPermission {
+		for _, permission := range response.CommunityChanges[0].TokenPermissionsAdded {
+			return permission
```
please, check that permission in a pending state

kounkou replied:
Solved with the new integration test.

mprakhov commented on protocol/communities_messenger_signers_test.go:
```diff
@@ -732,3 +732,127 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
 }
+
+func (s *MessengerCommunitiesSignersSuite) TestApplyendingCommunityEventsUponMakingDeviceControlNode() {
+	johnsOtherDevice := s.john
+	PairDevices(&s.Suite, johnsOtherDevice, s.john)
```
could you please check that `johnsOtherDevice` is not a control node?

kounkou replied:
This is not relevant anymore

mprakhov commented on protocol/messenger_communities.go:
```diff
@@ -4487,7 +4499,30 @@ func (m *Messenger) processCommunityChanges(messageState *ReceivedMessageState)
+	}
+
+	return false, nil
+}
```
please use `HasTokenOwnership` instead

kounkou replied:
Updated, good to know. thanks

kounkou committed:
fix_: mitigate permission stuck in pending state

This PR mitigates permission stuck in pending state upon making device a
control node. It fixes [#14023](status-im/status-desktop#14023)

MishkaRogachev commented on protocol/messenger_communities.go:
```diff
@@ -71,6 +71,18 @@ const (
 	maxChunkSizeBytes    = 1500000
 )
 
+const (
```
Kudos on keeping all the errors in one place!

MishkaRogachev reviewed (approved):
Changes look really solid!
